geometry/meshop.cpp

```diff
@@ -390,6 +502,162 @@ Mesh loadObj(const boost::filesystem::path &filename, ObjMaterial *mtl)
     return parser_.mesh;
 }
 
+/// Check that the following lines contain properties in the given order
+bool hasProperties(std::ifstream& f,
+                   std::string& line,
+                   const std::vector<std::string>& props)
+{
+    for (std::size_t i = 0; i < props.size(); ++i)
+    {
+        char pType[20];
+        char prop[20];
+        sscanf(line.c_str(), "property %s %s", pType, prop);
+
+        if (props[i] != prop) { return false; }
+        if (getline(f, line).eof()) { return false; }
+        boost::algorithm::trim(line);
+    }
+
+    return true;
+}
+
+Mesh loadPlyWithFeatures(const boost::filesystem::path& filename,
+                         math::Points3i* vertexColors,
+                         math::Points3i* faceColors,
+                         math::Points3* vertexNormals,
+                         std::vector<int>* faceLabels)
+{
+    std::ifstream f(filename.native());
+    if (!f.good())
+    {
+        LOGTHROW(err2, std::runtime_error) << "Can't open " << filename;
+    }
+
+    f.exceptions(std::ios::badbit | std::ios::failbit);
+
+    // read header
+    std::string line;
+
+    bool hasVertexColors = false;
+    bool hasAlphaVertexColor = false;
+    bool hasVertexNormals = false;
+    bool hasFaceColors = false;
+    bool hasAlphaFaceColor = false;
+    bool hasFaceLabels = false;
+    int nvert = -1;
+    int ntris = -1;
+
+    do
+    {
+        if (getline(f, line).eof()) { break; }
+        boost::algorithm::trim(line);
+        if (sscanf(line.c_str(), "element vertex %d", &nvert))
+        {
+            if (getline(f, line).eof()) { break; }
+            boost::algorithm::trim(line);
+            utility::expect(hasProperties(f, line, { "x", "y", "z" }),
```

Comment:
Does this work? Isn't it one property per line?

Reply:
Yes, the function `hasProperties()` is able to check multiple properties

Reply:
I see; the function reads additional lines inside.